### PR TITLE
[GEARPUMP-317] Fix Task minClock

### PR DESCRIPTION
**Pull request by manuzhang**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the commit message is formatted like:
   `[GEARPUMP-<Jira issue #>] Meaningful description of pull request` 
 - [x] Make sure tests pass via `sbt clean test`.
 - [x] Make sure old documentation affected by the pull request has been updated and new documentation added for new functionality.

### CODE REVIEW

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/task/GroupByTask.scala [hunk 1]
```diff
@@ -61,13 +62,14 @@ class GroupByTask[IN, GROUP, OUT](
   }
 
   override def onWatermarkProgress(watermark: Instant): Unit = {
-    groups.values.forEach(new Consumer[WindowRunner[IN, OUT]] {
-      override def accept(runner: WindowRunner[IN, OUT]): Unit = {
-        runner.trigger(watermark).foreach {
-          result =>
-            taskContext.output(Message(result.value, result.timestamp))
+    if (groups.isEmpty && watermark == Watermark.MAX) {
```
Comment (huafengw) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/task/GroupByTask.scala]: Why we have a special case here?

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/window/impl/WindowRunner.scala [hunk 2]
```diff
@@ -118,25 +126,34 @@ class DefaultWindowRunner[IN, OUT](
...
+            out: OUT =>
+              outputs += TimestampedValue(out, firstWin.endTime.minusMillis(1))
           }
+          val newWmk = TaskUtil.max(wmk, firstWin.endTime)
           if (windows.accumulationMode == Discarding) {
             fnRunner.teardown()
             setup = false
             // discarding, setup need to be called for each window
-            onTrigger(outputs)
+            onTrigger(outputs, newWmk)
```
Comment (huafengw) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/window/impl/WindowRunner.scala]: Move `onTrigger` outside of the `if else`

File: streaming/src/main/scala/org/apache/gearpump/streaming/task/Subscription.scala [hunk 3]
```diff
@@ -40,7 +41,7 @@ class Subscription(
     executorId: Int,
     taskId: TaskId,
     subscriber: Subscriber, sessionId: Int,
-    transport: ExpressTransport,
+    sender: TaskActor,
```
Comment (huafengw) [inline streaming/src/main/scala/org/apache/gearpump/streaming/task/Subscription.scala]: I think here `sender` is just the current task actor, while literally sender means the upstream task actor

File: streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala [hunk 4]
```diff
@@ -35,7 +35,7 @@ import org.apache.gearpump.streaming.ExecutorToAppMaster._
 import org.apache.gearpump.streaming.ProcessorId
 import org.apache.gearpump.streaming.task.TaskActor._
 import org.apache.gearpump.util.{LogUtil, TimeOutScheduler}
-import org.apache.gearpump.{MAX_TIME_MILLIS, MIN_TIME_MILLIS, Message, TimeStamp}
+import org.apache.gearpump.{Message, TimeStamp}
```
Comment (huafengw) [inline streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala]: Better not refactor the whole class next time...

**Comment by huafengw** [comment]
Generally +1

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/task/GroupByTask.scala (hunk 1 above)
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/task/GroupByTask.scala]: Other the `processingWatermark` will never advance when there are no incoming messages and `groups.isEmpty`

File: streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala (hunk 4 above)
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala]: Sure, I will bear it in mind.

**Pull request closed without merging**